File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s.length() == 0) return 0;
        int max = Integer.MIN_VALUE;
        int i = 0;
        int j = 0;
        HashSet<Character> set = new HashSet();
        while(i < s.length()) {
            if(!set.contains(s.charAt(i))) {
                set.add(s.charAt(i));
                max = Math.max(max, i-j+1);
                i++;
            }
            else if(set.contains(s.charAt(i))) {
                set.remove(s.charAt(j));
                j++;
            }
        }
        return max;
    }
}

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        #if len(s) == 0: return 0
        ans = 0
        left = 0
        right = 0
        
        hs = set()
        
        while(right < len(s)):
            if s[right] in hs:
                hs.remove(s[left])
                left += 1
            else:
                hs.add(s[right])
                ans = max(ans, right-left+1)
                right += 1
        
        return ans